File: 145/solution.clj
(defn num-to-digit-seq [n]
  (if (< n 10)
    (list n)
    (concat (num-to-digit-seq (quot n 10))
            (list (rem n 10)))))

(defn digit-seq-to-num
  ([xs] (digit-seq-to-num 0 xs))
  ([n xs]
    (if (= 0 (count xs)) n
      (digit-seq-to-num (+ (* n 10) (first xs)) (rest xs)))))

(defn reverse-num [n]
  (digit-seq-to-num (reverse (num-to-digit-seq n))))

(defn euler-reversible? [#^Integer n]
  (and
    (> (mod n 10) 0)
    (every? odd? (num-to-digit-seq (+ n (reverse-num n))))))

(def solution 
  (count (filter euler-reversible? (range 10000000 100000000))))

(print solution "\n")

; 100 -           20  2 digits - 20
; 1000 -         120  3 digits - 100
; 10000 -        720  4 digits - 600
; 100000 -       720  5 digits - ZERO
; 1000000 -    18720  6 digits - 18000
; 10000000 -   68720  7 digits - 50000
; 100000000 -         8 digits -
; 1000000000 -        9 digits - 
; 
; 1 3 5 7 9
; 

File: 145/solution.py
#!/usr/bin/env pypy

def digits(n):
    while n > 0:
        yield n % 10
        n /= 10

def trailing_zero(n):
    return n % 10 == 0

def reverse_digits(n):
    result = 0
    for digit in digits(n):
        result *= 10
        result += digit
    return result

def odd(n):
    return n % 2 == 1

def all_digits_odd(n):
    return all(odd(digit) for digit in digits(n))

def reversible(n):
    return not trailing_zero(n) and all_digits_odd(n + reverse_digits(n))

def solution(n):
    result = 0
    for i in xrange(1, n):
      if reversible(i):
        result += 1
    return result

if __name__ == "__main__":
    print solution(1000000000)
